hnn_core/network.py

```diff
@@ -764,136 +761,43 @@ def _attach_drive(self, name, drive, weights_ampa, weights_nmda, location,
-                drive_conn['src_gids'] = range(
-                    src_gid_curr,
-                    src_gid_curr + len(drive_conn['target_gids']))
-                src_gid_curr += len(drive_conn['target_gids'])
-                drive_conn_by_cell[cellname] = drive_conn
-        else:
-            drive_conn = dict()
-            drive_conn['location'] = location
-
-            # NB list! This is used later in _parnet_connect
-            drive_conn['src_gids'] = range(src_gid_curr,
-                                           src_gid_curr + n_drive_cells)
-            src_gid_curr += n_drive_cells
-            drive_conn['target_gids'] = list()  # fill in below
-            for cellname in target_populations:
-                drive_conn['target_gids'] = self.gid_ranges[cellname]
-                drive_conn['target_type'] = cellname
-                drive_conn_by_cell[cellname] = drive_conn.copy()
-
-        for cellname in target_populations:
-            for receptor, weights in weights_by_receptor.items():
-                drive_conn_by_cell[cellname][receptor] = dict()
-                if cellname in weights:
-                    drive_conn_by_cell[cellname][receptor][
-                        'lamtha'] = space_constant
-                    if isinstance(synaptic_delays, float):
-                        drive_conn_by_cell[cellname][receptor][
-                            'A_delay'] = synaptic_delays
-                    elif isinstance(synaptic_delays, dict):
-                        drive_conn_by_cell[cellname][receptor][
-                            'A_delay'] = synaptic_delays[cellname]
-                    drive_conn_by_cell[cellname][receptor][
-                        'A_weight'] = weights[cellname]
-
-        return drive_conn_by_cell, range(self._n_gids, src_gid_curr)
+        # Set the starting index for cell-specific source gids
+        # This will be updated depending on the number of target cells
+        # of each cell type
+        src_idx = 0
+        for target_cell_type in target_populations:
+            target_gids = list(self.gid_ranges[target_cell_type])
+            delays = delays_by_type[target_cell_type]
+            if cell_specific:
+                target_gids_nested = [[target_gid] for
+                                      target_gid in target_gids]
+                src_idx_end = src_idx + len(target_gids)
+                src_gids = (list(self.gid_ranges[name])
+                            [src_idx:src_idx_end])
+                src_idx = src_idx_end
+                for receptor in weights_by_type[target_cell_type]:
+                    weights = weights_by_type[target_cell_type][receptor]
+                    self.add_connection(src_gids, target_gids_nested,
+                                        location, receptor, weights, delays,
+                                        space_constant)
+            else:
+                src_gids = list(self.gid_ranges[name])
```

Comment:
```suggestion
```

shouldn't need this, no? You can just provide `name` to `add_connection` ?

Reply:
That only works for names in `Network.cell_types`. We could modify this validation step in `add_connection()`?

Reply:
I think we should ... it should allow `cell_types` as well as `external_drives.keys()`